File: api/app/utils/perimetres/arbre.py
"""Périmètres — l'arbre en base, lu et parcouru au même endroit partout.

⚠️ **Ce module est la moitié « arbre » du paquet `app.utils.perimetres`.** Il ne
rend aucun libellé : le rendu vit dans `libelles.py`, qui importe celui-ci. La
dépendance est à sens unique, et elle doit le rester — l'arbre n'a pas à savoir
comment on l'affiche.

Le découpage date du 27/08/2026 : le fichier unique passait 500 lignes, seuil de
la règle de modularité (rang 1). La frontière suit celle que ce docstring
énonçait déjà — « Les primitives » ici, « Libellés » là-bas. La surface publique,
elle, N'A PAS BOUGÉ : `from app.utils.perimetres import …` continue de tout
trouver, via `__init__.py`.

## Ce que ce module remplaçait

Un périmètre (« résidence », « bat:3 », « parking ») était une **table de libellés
écrite en dur**, et elle l'était **trois fois** le 08/08/2026 :

- `app/routers/flux/commun.py` — table complète, bâtiments 1 à 9, AFUL ;
- `app/routers/tickets.py` (relance syndic) — table partielle : ni AFUL, ni
  bâtiment au-delà de ce que le préfixe `bat:` produisait ;
- `front/src/lib/utils.ts` — côté interface, arrêtée à `bat:4`.

Les tables Python ne donnaient pas le même libellé pour un même périmètre, et le
front en donnait un troisième : un ticket ciblé AFUL sortait « aful » dans l'e-mail
de relance, « AFUL » dans le fil, et un `bat:5` s'affichait « Bât. 5 » côté API mais
`bat:5` **brut** à l'écran. C'est la divergence typique décrite par
`standards/02-factorisation.md` §2.

## Ce qui gouverne ce module maintenant

**Aucun code de périmètre n'est écrit ici.** Le produit doit servir une autre
copropriété, qui n'a ni AFUL, ni quatre bâtiments, ni forcément de caves — et qui
aura peut-être une piscine et trois entrées. L'arbre vit dans la table `perimetre`,
il est reconstructible depuis l'administration, et **le vider ne casse rien** : un
périmètre vide vaut « concerne tout le monde », ce qui était déjà la règle
(`visibility.perimetre_visible`).

C'est pourquoi `parse_perimetres` ne renvoie plus la chaîne littérale
« résidence » quand le champ est vide, mais le **nœud racine à portée globale**
désigné par les données — voir `code_par_defaut`.

## Les primitives

`a_portee_globale`, `batiments_cibles` et `perimetre_du_batiment` sont les
seules fonctions qui parcourent l'arbre. Leurs
consommateurs — `utils/visibility.py` (qui voit), `utils/destinataires.py` (qui
est notifié), `routers/flux/evenements.py` (le badge « concerne mon bâtiment ») et
`utils/fiche_arrivant.py` (le document imprimé) — portaient auparavant chacun sa
propre copie de la liste des périmètres transverses, ou sa propre convention de
nommage.
"""
from __future__ import annotations

import json
import logging
import time
from typing import NamedTuple, Optional

logger = logging.getLogger("hostachy.perimetres")

#: Durée de vie du cache. Le cache est vidé explicitement à chaque écriture
#: (`invalider_cache`), mais uvicorn peut servir plusieurs *workers* : une
#: invalidation dans l'un n'atteint pas les autres. Ce délai garantit la
#: convergence sans bus de messages — un libellé corrigé dans l'administration
#: apparaît partout en moins d'une demi-minute.
_TTL_SECONDES = 30.0

_PREFIXE_BATIMENT = "bat:"


class Noeud(NamedTuple):
    """Un périmètre, tel que l'arbre le donne. Immuable, sans session attachée."""
    code: str
    libelle: str
    libelle_court: str
    description: str
    icone: Optional[str]
    parent: Optional[str]
    batiment_id: Optional[int]
    portee_globale: bool
    selectionnable: bool
    actif: bool
    ordre: int


_cache: Optional[dict[str, Noeud]] = None
_cache_pose: float = 0.0


def invalider_cache() -> None:
    """À appeler après toute écriture dans la table `perimetre`."""
    global _cache, _cache_pose
    _cache = None
    _cache_pose = 0.0


def arbre() -> dict[str, Noeud]:
    """L'arbre entier, indexé par code **en minuscules**.

    Renvoie un dictionnaire vide si la table n'existe pas encore (premier
    démarrage, avant `create_db_and_tables`) ou si elle est vide — et c'est un état
    valide, pas une erreur : une copropriété qui n'a pas encore configuré ses
    périmètres n'en restreint aucun.
    """
    global _cache, _cache_pose
    if _cache is not None and (time.monotonic() - _cache_pose) < _TTL_SECONDES:
        return _cache

    #  Imports différés : ce module est importé par `utils/visibility.py`, lui-même
    #  importé très tôt. Charger la base au moment de l'appel, pas de l'import.
    try:
        from sqlmodel import select

        from app.database import SessionLocal
        from app.models.perimetre import Perimetre

        with SessionLocal() as session:
            lignes = session.exec(select(Perimetre)).all()
            #  Le parent est exposé par son CODE et non son id : tout le reste du
            #  produit raisonne en codes, et un cache qui mélangerait les deux
            #  obligerait chaque appelant à faire la conversion.
            code_par_id = {ligne.id: ligne.code.lower() for ligne in lignes}
            construit = {
                ligne.code.lower(): Noeud(
                    code=ligne.code,
                    libelle=ligne.libelle,
                    libelle_court=ligne.libelle_court or ligne.libelle,
                    description=ligne.description or "",
                    icone=ligne.icone,
                    parent=code_par_id.get(ligne.parent_id),
                    batiment_id=ligne.batiment_id,
                    portee_globale=bool(ligne.portee_globale),
                    selectionnable=bool(ligne.selectionnable),
                    actif=bool(ligne.actif),
                    ordre=ligne.ordre or 0,
                )
                for ligne in lignes
            }
    except Exception as exc:
        #  Table absente ou base indisponible. On ne met **pas** le résultat en
        #  cache : la table apparaîtra au prochain démarrage, et figer un arbre
        #  vide pour 30 s masquerait le rétablissement.
        #
        #  Conséquence à connaître, et c'est pourquoi ce journal est en `error` :
        #  un arbre illisible ne permet plus de résoudre un périmètre, donc les
        #  contenus qui en citent un deviennent invisibles des résidents (le
        #  conseil syndical et l'administration continuent de tout voir). C'est le
        #  sens choisi : un contrôle qui ne peut pas s'exécuter ne renvoie pas OK
        #  (`standards/04`). L'inverse aurait rendu lisible, sur un simple hoquet
        #  de base, un document réservé à un autre bâtiment.
        logger.error(
            "Arbre des périmètres illisible (%s) — les contenus à périmètre "
            "deviennent invisibles des résidents jusqu'au rétablissement", exc,
        )
        return {}

    _cache = construit
    _cache_pose = time.monotonic()
    return _cache


# ── Parcours de l'arbre : les trois primitives ────────────────────────────────

def _chaine(code: str, noeuds: dict[str, Noeud]) -> list[Noeud]:
    """Le nœud puis ses ancêtres, du plus proche au plus lointain.

    Renvoie une liste **vide** si le code est inconnu : un contenu qui cite un
    périmètre supprimé n'accorde aucun droit — il n'en retire pas non plus, il ne
    compte simplement pas. Un `parent_id` orphelin ou un cycle arrête la remontée
    sans lever : `visited` est le garde-fou, parce qu'une boucle dans l'arbre ne
    doit jamais pouvoir suspendre une requête (`standards/04` — un contrôle qui ne
    peut pas s'exécuter ne renvoie pas OK).
    """
    depart = noeuds.get(code.strip().lower())
    if depart is None:
        return []
    chaine = [depart]
    vus = {depart.code.lower()}
    courant = depart
    while courant.parent:
        suivant = noeuds.get(courant.parent)
        if suivant is None or suivant.code.lower() in vus:
            break
        chaine.append(suivant)
        vus.add(suivant.code.lower())
        courant = suivant
    return chaine


def a_portee_globale(codes: list[str]) -> bool:
    """L'un de ces périmètres concerne-t-il tous les résidents ?

    Vrai dès qu'un nœud cité — ou l'un de ses ancêtres — porte `portee_globale`.
    L'héritage est ce qui permet de cibler « Parking › Portail d'accès » sans avoir
    à rappeler que le parking concerne tout le monde.
    """
    noeuds = arbre()
    if not noeuds:
        return False
    return any(
        n.portee_globale
        for code in codes
        for n in _chaine(code, noeuds)
    )


def batiments_cibles(codes: list[str]) -> set[int]:
    """Les identifiants de bâtiments réellement visés par ces périmètres.

    Pour chaque code, on retient le `batiment_id` du nœud ou, à défaut, celui du
    plus proche ancêtre qui en porte un — c'est ainsi que « Bât. 2 › Hall d'entrée »
    concerne le bâtiment 2 sans le répéter sur chaque espace.
    """
    noeuds = arbre()
    if not noeuds:
        return set()
    cibles: set[int] = set()
    for code in codes:
        for n in _chaine(code, noeuds):
            if n.batiment_id is not None:
                cibles.add(n.batiment_id)
                break
    return cibles


def perimetre_du_batiment(batiment_id: Optional[int]) -> Optional[Noeud]:
    """Le nœud de l'arbre qui **est** ce bâtiment — chemin inverse de `batiments_cibles`.

    Sert à tout ce qui part d'un bâtiment (un membre du conseil syndical, un lot)
    et doit le **nommer** : le nom et l'icône viennent alors de l'arbre, donc de
    l'administration, au lieu d'être fabriqués par un `f"Bât. {numero}"` recopié.
    Cette convention-là existait dans sept fichiers le 14/08/2026 ; le document
    imprimé était le seul endroit où elle produisait un libellé qu'aucun
    renommage ne rattrapait, puisqu'il ne passe par aucun écran.

    Renvoie `None` si l'arbre est vide, si le bâtiment n'y figure pas, ou si son
    nœud est inactif — l'appelant garde alors son propre repli. Ne jamais faire
    lever : un document doit se produire même sur un arbre incomplet.

    Seul le nœud « bâtiment » porte un `batiment_id` ; ses espaces (hall,
    ascenseur…) l'héritent de leur ancêtre, ce que `batiments_cibles` exploite.
    Si plusieurs nœuds actifs le portaient malgré tout — données reprises à la
    main — on retient le plus prioritaire, jamais un au hasard.
    """
    if batiment_id is None:
        return None
    candidats = [
        n for n in arbre().values()
        if n.batiment_id == batiment_id and n.actif
    ]
    if not candidats:
        return None
    return min(candidats, key=lambda n: (n.ordre, n.code))


# ── Analyse des champs stockés ────────────────────────────────────────────────

def code_par_defaut() -> Optional[str]:
    """Le périmètre qu'un contenu sans périmètre explicite désigne implicitement.

    C'est la racine à portée globale la plus prioritaire — « Copropriété entière »
    sur l'arbre livré. Cette fonction existe pour que la valeur par défaut soit une
    **donnée** et non la chaîne « résidence » écrite dans le code : une
    copropriété qui renomme ou supprime ce nœud ne doit pas casser l'application.

    Renvoie `None` sur un arbre vide, et alors `parse_perimetres` rend une liste
    vide — que `perimetre_visible` traite déjà comme « visible de tous ».
    """
    candidats = [
        n for n in arbre().values()
        if n.portee_globale and n.parent is None and n.actif
    ]
    if not candidats:
        return None
    return min(candidats, key=lambda n: (n.ordre, n.code)).code


def _avec_defaut(codes: list[str]) -> list[str]:
    if codes:
        return codes
    defaut = code_par_defaut()
    return [defaut] if defaut else []


def parse_perimetres(perimetre: Optional[str]) -> list[str]:
    """Champ `perimetre` en texte (« résidence », « parking,cave »)."""
    if not perimetre:
        return _avec_defaut([])
    return _avec_defaut([s.strip() for s in perimetre.split(",") if s.strip()])


def parse_json_perimetres(perimetre_cible: Optional[str]) -> list[str]:
    """Champ `perimetre_cible` en JSON (ex. '["bat:1","bat:3"]').

    Un contenu illisible retombe sur le périmètre par défaut plutôt que de lever :
    ce champ alimente un affichage, il ne doit jamais faire échouer une requête.

    ⚠️ Ce repli est **volontairement limité à l'affichage**. Pour décider d'un
    accès, `visibility.perimetre_visible` n'utilise pas ce repli : un JSON corrompu
    ne doit pas ouvrir une visibilité, alors qu'il peut bien rendre un badge
    générique.
    """
    if not perimetre_cible:
        return _avec_defaut([])
    try:
        val = json.loads(perimetre_cible) if isinstance(perimetre_cible, str) else perimetre_cible
        codes = [str(v) for v in val] if isinstance(val, (list, tuple)) else []
    except Exception:
        codes = []
    return _avec_defaut(codes)

File: front/src/lib/perimetres/arbre.ts
import { perimetreLabel, perimetreLabelUn } from './libelles';

/**
 * Périmètres — le rendu, sans table.
 *
 * ## Ce que ce module remplace
 *
 * `lib/utils.ts` portait `PERIMETRE_LABELS`, une table de libellés écrite en dur
 * et arrêtée à `bat:4`, quand l'API allait jusqu'à `bat:9` : un cinquième bâtiment
 * s'affichait « Bât. 5 » côté serveur et **`bat:5` brut** à l'écran. Trois copies
 * s'étaient installées autour (#316), toutes correctes, toutes divergentes à terme.
 *
 * L'arborescence vit désormais en base et s'édite depuis l'administration. Ce
 * module n'en garde qu'un **cache d'affichage**, rempli par
 * `stores/perimetres.ts` au démarrage.
 *
 * ## Pourquoi un module à part, sans aucun import
 *
 * `perimetreLabel()` est appelée dans une quinzaine de gabarits, en plein milieu
 * du rendu : elle doit rester **synchrone**. La rendre asynchrone aurait imposé de
 * réécrire chacun de ces appels. Elle lit donc une carte de module, et le store
 * s'occupe de la remplir — c'est le store qui connaît l'API, pas l'inverse.
 *
 * N'importer rien ici est délibéré : `lib/utils.ts` réexporte `perimetreLabel`, et
 * une dépendance vers `$lib/api` créerait un cycle avec les modules qui importent
 * `utils`.
 */

/** Un nœud de l'arborescence, tel que `GET /perimetres` le rend. */
export interface Perimetre {
	id: number;
	code: string;
	parent: string | null;
	libelle: string;
	libelle_court: string;
	description: string;
	icone: string | null;
	batiment_id: number | null;
	profondeur: number;
	ordre: number;
	actif: boolean;
	portee_globale: boolean;
	concerne_tous: boolean;
	selectionnable: boolean;
	/**  Cet espace est-il PRIVATIF — un logement, une cave, une place attribuée —
	 *   par opposition aux parties communes ? Administré depuis Admin →
	 *   Patrimoine, jamais deviné d'après le libellé : chaque copropriété nomme
	 *   ses espaces comme elle veut (31/08/2026).
	 *
	 *   ⚠️ **Purement visuel pour l'instant.** La pastille se distingue, rien
	 *   d'autre ne change — ni qui voit, ni qui est notifié. */
	privatif: boolean;
	utilise: boolean;
}

export const PREFIXE_BATIMENT = 'bat:';

//  Exportée pour `libelles.ts` seul : le paquet la partage, le dépôt ne la
//  voit pas — `index.ts` ne la réexporte pas.
export let carte: Record<string, Perimetre> = {};
let ordreCodes: string[] = [];
let codeDefaut: string | null = null;

export const cle = (code: string) => (code ?? '').trim().toLowerCase();

/**
 * Remplit le cache d'affichage. Appelé par `stores/perimetres.ts`, nulle part ailleurs.
 */
export function definirPerimetres(liste: Perimetre[]): void {
	carte = Object.fromEntries(liste.map((n) => [cle(n.code), n]));
	ordreCodes = liste.map((n) => n.code);
	//  Le périmètre « tout le monde » est une DONNÉE, pas la chaîne « résidence »
	//  écrite dans le code : une autre copropriété peut l'avoir renommé ou supprimé.
	//  C'est la racine à portée globale la plus prioritaire, comme côté serveur
	//  (`api/app/utils/perimetres.py`, `code_par_defaut`).
	const racines = liste.filter((n) => n.parent === null && n.portee_globale && n.actif);
	codeDefaut = racines.length ? racines[0].code : null;
}

/** L'arborescence connue, dans l'ordre d'affichage rendu par l'API. */
export function tousLesPerimetres(): Perimetre[] {
	return ordreCodes.map((c) => carte[cle(c)]).filter(Boolean);
}

export function noeudPerimetre(code: string): Perimetre | undefined {
	return carte[cle(code)];
}

/**
 * Le code du périmètre qui représente un bâtiment donné.
 *
 * Six endroits du front construisaient `` `bat:${batiment_id}` `` à la main —
 * la convention de nommage du seed recopiée en dur, alors que l'administration
 * peut créer un bâtiment sous n'importe quel code. On interroge l'arbre.
 *
 * Retombe sur la convention si l'arborescence n'est pas encore chargée : c'est un
 * affichage, et le libellé se corrigera au chargement.
 */
export function perimetreDuBatiment(batimentId: number | null | undefined): string {
	if (batimentId === null || batimentId === undefined) return perimetreParDefaut() ?? '';
	const trouve = tousLesPerimetres().find((n) => n.batiment_id === batimentId && n.actif);
	return trouve ? trouve.code : `${PREFIXE_BATIMENT}${batimentId}`;
}

/** Le code qui désigne « toute la copropriété », ou `null` sur un arbre vide. */
export function perimetreParDefaut(): string | null {
	return codeDefaut;
}

/**
 * La sélection initiale d'un formulaire — « toute la copropriété ».
 *
 * Les pages écrivaient `['résidence']` en dur, une trentaine de fois. Sur une
 * copropriété qui renomme ou supprime ce nœud, chacune de ces occurrences aurait
 * produit un périmètre inexistant, sans erreur visible : le formulaire se serait
 * ouvert sur une pastille morte. Liste vide sur un arbre vide, ce que le serveur
 * traite déjà comme « concerne tout le monde ».
 */
export function perimetreDefautListe(): string[] {
	return codeDefaut ? [codeDefaut] : [];
}

/**
 * Cette sélection désigne-t-elle toute la copropriété ?
 *
 * Remplace les comparaisons `=== 'résidence'` semées dans les pages, qui
 * cesseraient d'être vraies dès qu'une copropriété renomme ou supprime ce nœud.
 * Une liste vide vaut « tout le monde », comme côté serveur.
 */
export function estPerimetreParDefaut(items: string[] | string | null | undefined): boolean {
	const liste = normaliser(items);
	if (liste.length === 0) return true;
	return liste.length === 1 && codeDefaut !== null && cle(liste[0]) === cle(codeDefaut);
}

export function normaliser(items: string[] | string | null | undefined): string[] {
	const liste = typeof items === 'string' ? items.split(',') : (items ?? []);
	return liste.map((i) => (i ?? '').trim()).filter(Boolean);
}

/**
 * L'un de ces périmètres concerne-t-il tous les résidents ?
 *
 * Miroir de `a_portee_globale` côté serveur. L'héritage est déjà résolu par l'API,
 * qui rend `concerne_tous` pour chaque nœud — inutile de remonter l'arbre ici.
 *
 * Remplace la liste `['résidence','parking','cave','aful']` que le tableau de bord
 * portait en dur, troisième copie d'une même énumération.
 */
export function concerneTous(items: string[] | string | null | undefined): boolean {
	return normaliser(items).some((c) => carte[cle(c)]?.concerne_tous === true);
}

/**
 * Les bâtiments réellement visés — celui du nœud, ou du plus proche ancêtre qui en
 * porte un. C'est ce qui fait que « Bât. 2 › Hall d'entrée » concerne le bâtiment 2
 * sans que le hall ait à le répéter.
 *
 * Miroir de `batiments_cibles` côté serveur. La remontée est bornée par `vus` : une
 * boucle dans l'arbre ne doit pas figer l'interface.
 */
export function batimentsCibles(items: string[] | string | null | undefined): number[] {
	const cibles = new Set<number>();
	for (const code of normaliser(items)) {
		let courant = carte[cle(code)];
		const vus = new Set<string>();
		while (courant && !vus.has(cle(courant.code))) {
			if (courant.batiment_id !== null) {
				cibles.add(courant.batiment_id);
				break;
			}
			vus.add(cle(courant.code));
			courant = courant.parent ? carte[cle(courant.parent)] : undefined!;
		}
	}
	return [...cibles];
}

/**
 * Icônes proposées pour un périmètre, dans l'écran d'administration.
 *
 * Volontairement **courte et thématique** : un choix de cinquante icônes ne se
 * parcourt pas, il se subit. Chacune doit dire quelque chose d'un lieu de
 * copropriété — et toutes existent dans `Icon.svelte`, sans quoi la pastille
 * s'afficherait avec le point d'interrogation du repli.
 */
/**
 * Les périmètres de PREMIER NIVEAU — ceux qui font une rangée de pastilles.
 *
 * Un nœud de premier niveau est soit une **racine sélectionnable**, soit
 * l'**enfant d'un regroupement racine** : « Bâtiments » n'est pas une cible, on
 * choisit un bâtiment. Cela remonte les bâtiments dans la première rangée, là où
 * on les cherche, *sans inventer de niveau dans les données*.
 *
 * 🔴 Cette règle vivait dans `PerimetrePicker.svelte` jusqu'au 10/09/2026, où le
 * carnet d'entretien a eu besoin de la même rangée **pour filtrer**. Un filtre
 * qui l'aurait recalculée à sa façon aurait divergé au premier périmètre créé en
 * administration — et c'est justement ce que ce filtre doit savoir faire : suivre
 * l'arborescence sans qu'on y touche.
 *
 * @param exclure — le code à ne pas rendre. Le sélecteur écarte le périmètre
 *   PAR DÉFAUT (il est déjà la valeur initiale) ; un filtre l'écarte aussi, mais
 *   le remplace par sa propre pastille « tout ». Deux usages, une règle.
 */
export function perimetresNiveau1(liste: Perimetre[], exclure?: string | null): Perimetre[] {
	const actifs = liste.filter((n) => n.actif);
	const parCode = new Map(actifs.map((n) => [n.code, n]));
	const estGroupeRacine = (n: Perimetre | undefined): boolean =>
		!!n && n.parent === null && !n.selectionnable;
	return actifs.filter(
		(n) =>
			n.selectionnable &&
			n.code !== exclure &&
			(n.parent === null || estGroupeRacine(parCode.get(n.parent!))),
	);
}

/**
 * Le périmètre dont HÉRITE une nouvelle entrée d'historique.
 *
 * 🔴 Signalé à l'écran le 31/08/2026 :
 *
 * > *« quand on fait un commentaire sur un Ticket, par défaut le périmètre du
 * > dernier commentaire (ou du ticket original si 1er commentaire) n'est pas
 * > conservé »*
 *
 * Le formulaire proposait « Copropriété entière » sur un ticket situé « Bât. 1 ›
 * Escaliers ». Il ne mentait pas sur ce qui allait s'écrire — le serveur ne
 * touche à rien quand le champ est vide — mais il **montrait un choix par défaut
 * qui n'était pas celui qui s'appliquerait**, ce qui revient au même pour qui
 * lit l'écran.
 *
 * ⚠️ L'héritage remonte les entrées, il ne prend pas seulement le ticket : une
 * entrée a pu resserrer le périmètre — « on a trouvé d'où vient la fuite » — et
 * c'est ce resserrement, le plus récent, qui vaut ensuite. Prendre le ticket
 * ferait revenir en arrière à chaque commentaire.
 *
 * ⚠️ Une entrée qui ne dit RIEN du périmètre ne compte pas : elle n'a rien
 * précisé, donc elle n'a rien changé. C'est le sens de la valeur vide, et c'est
 * ce qui permet à un courriel d'afficher « 🔹 … » sur une entrée pour dire
 * qu'elle a précisé quelque chose.
 *
 * @param perimetreObjet le périmètre de l'objet porteur (ticket, événement…)
 * @param entrees        l'historique, dans l'ordre CHRONOLOGIQUE
 */
export function perimetreHerite(
	perimetreObjet: string[] | null | undefined,
	entrees: { perimetre_cible?: string[] | null }[] = [],
): string[] {
	for (let i = entrees.length - 1; i >= 0; i--) {
		const precise = entrees[i]?.perimetre_cible;
		if (precise && precise.length) return [...precise];
	}
	return [...(perimetreObjet ?? [])];
}

/**
 * Deux périmètres désignent-ils la même chose ? **L'ORDRE ne compte pas.**
 *
 * Le sélecteur mémorise l'ordre des clics ; deux mêmes zones cochées dans un
 * autre ordre sont le même périmètre. C'est déjà la règle de `perimetreLabel`,
 * qui trie avant de rendre — la comparer autrement ferait diverger l'affichage
 * et la décision.
 *
 * ⚠️ Employé pour n'envoyer `perimetre_cible` que s'il DIFFÈRE de l'hérité : une
 * comparaison sensible à l'ordre ferait déclarer un resserrement à chaque
 * commentaire où l'on aurait décoché puis recoché la même zone.
 */
export function memePerimetre(a: string[], b: string[]): boolean {
	return a.length === b.length && [...a].sort().join('|') === [...b].sort().join('|');
}

/**
 * Le périmètre d'une nouvelle entrée d'historique : ce qu'elle déclare, et le
 * libellé de ce dont elle part.
 *
 * ## Pourquoi une fonction et pas trois `$:` dans le formulaire
 *
 * Les deux valeurs se répondent, et la subtile est `declare` : elle vaut
 * `undefined` tant qu'on n'a pas touché à l'hérité. C'est ce **vide** qui permet
 * au courriel de n'afficher « 🔹 … » que sur les entrées qui ont précisé quelque
 * chose — une entrée qui redit le périmètre du ticket n'apprend rien.
 *
 * Elles vivaient dans `EvolForm`, qui sert quatre écrans et qui a franchi le
 * plafond de modularité le 31/08/2026. Le contrôle désignait un **placement** :
 * ces trois lignes parlent de périmètre, pas de formulaire, et leur place est
 * ici — à côté de `perimetreHerite`, dont elles sont les seules clientes.
 */
export function perimetreEntree(
	perimetreCourant: string[],
	entrees: { perimetre_cible?: string[] | null }[],
	choisi: string[],
	sectionActive: boolean,
): { declare: string[] | undefined; libelleActuel: string } {
	const depart = perimetreHerite(perimetreCourant, entrees);
	return {
		declare: sectionActive && choisi.length && !memePerimetre(choisi, depart) ? choisi : undefined,
		libelleActuel: depart.length
			? perimetreLabel(depart)
			: perimetreLabelUn(perimetreParDefaut() ?? ''),
	};
}
